src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractor.java

```diff
@@ -0,0 +1,241 @@
+
+  private static final String SUPPORT = "support";
+
+  private static final String SUPPORT_EMAIL = "email";
+
+  private static final String SUPPORT_ISSUES = "issues";
+
+  private static final String SUPPORT_FORUM = "forum";
+
+  private static final String SUPPORT_WIKI = "wiki";
+
+  private static final String SUPPORT_IRC = "irc";
+
+  private static final String SUPPORT_SOURCE = "source";
+
+  private static final String SUPPORT_DOCS = "docs";
+
+  private static final String SUPPORT_RSS = "rss";
+
+  private static final Map<String, String> STRINGS_MAPPING = new ImmutableMap.Builder<String, String>()
+      .put(NAME, P_NAME)
+      .put(DESCRIPTION, P_DESCRIPTION)
+      .put(VERSION, P_VERSION)
+      .put(TYPE, P_TYPE)
+      .put(KEYWORDS, P_KEYWORDS)
+      .put(HOMEPAGE, P_HOMEPAGE)
+      .put(TIME, P_TIME)
+      .put(LICENSE, P_LICENSE)
+      .build();
+
+  private static final Map<String, String> SUPPORT_MAPPING = new ImmutableMap.Builder<String, String>()
+      .put(SUPPORT_EMAIL, P_SUPPORT_EMAIL)
+      .put(SUPPORT_ISSUES, P_SUPPORT_ISSUES)
+      .put(SUPPORT_FORUM, P_SUPPORT_FORUM)
+      .put(SUPPORT_WIKI, P_SUPPORT_WIKI)
+      .put(SUPPORT_IRC, P_SUPPORT_IRC)
+      .put(SUPPORT_SOURCE, P_SUPPORT_SOURCE)
+      .put(SUPPORT_DOCS, P_SUPPORT_DOCS)
+      .put(SUPPORT_RSS, P_SUPPORT_RSS)
+      .build();
+
+  private final TypeReference<Map<String, Object>> typeReference = new TypeReference<Map<String, Object>>() { };
+
+  private final ObjectMapper mapper = new ObjectMapper();
+
+  private final ArchiveStreamFactory archiveStreamFactory = new ArchiveStreamFactory();
+
+  /**
+   * Populates an asset's format attributes with the content contained in a composer.json file in the zip archive. This
+   * does not extract all JSON entries, but does try to extract those that could be viewed as more "interesting" from
+   * the standpoint of the repository manager.
+   */
+  public void extractFromZip(final TempBlob tempBlob, final NestedAttributesMap formatAttributes) throws IOException {
+    try (InputStream is = tempBlob.getBlob().getInputStream()) {
+      try (ArchiveInputStream ais = archiveStreamFactory.createArchiveInputStream(ArchiveStreamFactory.ZIP, is)) {
+        ArchiveEntry entry = ais.getNextEntry();
+        while (entry != null) {
+          if (processEntry(ais, entry, formatAttributes)) {
+            return;
+          }
+          entry = ais.getNextEntry();
+        }
+      }
+    }
+    catch (ArchiveException e) {
+      throw new IOException("Error reading from archive", e);
+    }
+  }
+
+  /**
+   * Processes a single entry in the archive. If the entry is the composer.json then the attributes will be extracted.
+   * If not, the entry is skipped.
+   */
+  private boolean processEntry(final ArchiveInputStream stream,
+                               final ArchiveEntry entry,
+                               final NestedAttributesMap formatAttributes) throws IOException
```

Comment:
Question: Is there a reason you chose to pass the AttributesMap rather than returning a built AttributesMap to be set into the Asset?

Reply:
> Question: Is there a reason you chose to pass the AttributesMap rather than returning a built AttributesMap to be set into the Asset?

Just that we were going to be setting all these into the formatAttributes portion of a NestedAttributesMap, and so you can't just "set" those values in (unless you're going to try and modify the backing map yourself, which seems to counter the intent behind the NestedAttributesMap in the first place)

I agree that it's inelegant to do it this way but it seemed no worse than returning a map and then having to iterate over all the keys to copy them into the destination formatAttributes NestedAttributesMap one by one, which is what we usually end up having to do in these situations. (Though if I were going to do this over, I might consider doing it that way since I do think the code ends up being more testable.)